services/s3/src/main/java/software/amazon/awssdk/services/s3/handlers/EndpointAddressInterceptor.java

```diff
@@ -45,28 +48,82 @@ public SdkHttpFullRequest modifyHttpRequest(Context.ModifyHttpRequest context, E
 
         S3AdvancedConfiguration advancedConfiguration =
                 (S3AdvancedConfiguration) executionAttributes.getAttribute(AwsExecutionAttributes.SERVICE_ADVANCED_CONFIG);
-
         SdkHttpFullRequest.Builder mutableRequest = request.toBuilder();
 
+        mutableRequest.endpoint(resolveEndpoint(request.getEndpoint(), originalRequest,
+                                                executionAttributes, advancedConfiguration));
+
         if (advancedConfiguration == null || !advancedConfiguration.pathStyleAccessEnabled()) {
             try {
                 String bucketName = getBucketName(originalRequest);
 
                 if (BucketUtils.isValidDnsBucketName(bucketName, false)) {
-                    changeToDnsEndpoint(request, mutableRequest, bucketName);
+                    changeToDnsEndpoint(mutableRequest, bucketName);
                 }
             } catch (Exception e) {
                 // Unable to convert to DNS style addressing. Fall back to continue using path style.
             }
         }
 
-        // Remove accelerate from operations that don't allow it
-        if (advancedConfiguration != null && advancedConfiguration.accelerateModeEnabled()
-            && ACCELERATE_DISABLED_OPERATIONS.contains(originalRequest.getClass())) {
-            removeAccelerate(request, mutableRequest);
+        return mutableRequest.build();
+    }
+
+    /**
+     * Determine which endpoint to use based on region and {@link S3AdvancedConfiguration}. Will either be a traditional
+     * S3 endpoint (i.e. s3.us-east-1.amazonaws.com), the global S3 accelerate endpoint (i.e. s3-accelerate.amazonaws.com) or
+     * a regional dualstack endpoint for IPV6 (i.e. s3.dualstack.us-east-1.amazonaws.com).
+     */
+    private URI resolveEndpoint(URI originalEndpoint,
+                                Object originalRequest,
+                                ExecutionAttributes executionAttributes,
+                                S3AdvancedConfiguration advancedConfiguration) {
+        Region region = executionAttributes.getAttribute(AwsExecutionAttributes.AWS_REGION);
+        RegionMetadata regionMetadata = RegionMetadata.of(region);
+        if (isAccelerateEnabled(advancedConfiguration) && isAccelerateSupported(originalRequest)) {
+            return accelerateEndpoint(advancedConfiguration, regionMetadata);
+        } else if (advancedConfiguration != null && advancedConfiguration.dualstackEnabled()) {
```

Comment:
Nit: we usually forego the `else` when the only statement in the if is a `return`

Reply:
Gotcha; great point!